### PR TITLE
Adds a proof harness for _IotMqtt_Notify function

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*
- Adds a new proof harness for the `_IotMqtt_Notify ` function;
- Updates `valid_IotMqttOperation` in MQTT state.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cbmc/proofs/IotMqtt_Notify/Makefile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# One more than actual number of operations
+OPERATION_COUNT_MAX=4
+DEF += -DOPERATION_COUNT_MAX=$(OPERATION_COUNT_MAX)
+
+# One more than actual length of topics
+TOPIC_LENGTH_MAX=10
+DEF += -DTOPIC_LENGTH_MAX=$(TOPIC_LENGTH_MAX)
+
+# packet size
+DEF += -DPACKET_SIZE_MAX=10
```
Comment (markrtuttle) [inline cbmc/proofs/IotMqtt_Notify/Makefile:26]: Where is packet size max used?

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+typedef bool ( * MatchFunction_t )( const IotLink_t * const pOperationLink,
+                                    void * pCompare );
+
+typedef void ( *FreeElementFunction_t )( void * pData );
+
+/**
+ * We constrain the return values of these functions because
+ * they are checked by assertions in the MQTT code.
+ */
```
Comment (markrtuttle) [inline cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:45]: These models need to change from proof to proof, are you sure this is the right one?

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    __CPROVER_assume( IS_STUBBED_NETWORKIF_SEND( pMqttConnection->pNetworkInterface ) );
+    __CPROVER_assume( IS_STUBBED_NETWORKIF_DESTROY( pMqttConnection->pNetworkInterface ) );
+    ensure_IotMqttConnection_has_lists( pMqttConnection );
+    __CPROVER_assume( valid_IotMqttConnection( pMqttConnection ) );
+    __CPROVER_assume( pMqttConnection->references > 0 );
+
+    /* Assume unconstrained operation. */
+    IotMqttOperation_t pOperation = allocate_IotMqttOperation( NULL, pMqttConnection );
+    __CPROVER_assume( valid_IotMqttOperation( pOperation ) );
+    __CPROVER_assume( pOperation->incomingPublish == false );
```
Comment (markrtuttle) [inline cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:106]: Is this true?  Give a comment saying why.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+{
+    /* Assume a valid MQTT connection. */
+    IotMqttConnection_t pMqttConnection = allocate_IotMqttConnection( NULL );
+    __CPROVER_assume( pMqttConnection != NULL );
+    __CPROVER_assume( pMqttConnection->pNetworkInterface != NULL );
+    __CPROVER_assume( IS_STUBBED_NETWORKIF_SEND( pMqttConnection->pNetworkInterface ) );
+    __CPROVER_assume( IS_STUBBED_NETWORKIF_DESTROY( pMqttConnection->pNetworkInterface ) );
+    ensure_IotMqttConnection_has_lists( pMqttConnection );
+    __CPROVER_assume( valid_IotMqttConnection( pMqttConnection ) );
+    __CPROVER_assume( pMqttConnection->references > 0 );
```
Comment (markrtuttle) [inline cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:101]: Is this true?  Give a comment saying why.

File: cbmc/proofs/mqtt_state.c [hunk 5]
```diff
@@ -110,12 +110,17 @@ bool valid_IotMqttOperation( const IotMqttOperation_t pOp )
   bool waitable =
     ( pOp->u.operation.flags & IOT_MQTT_FLAG_WAITABLE )
         == IOT_MQTT_FLAG_WAITABLE;
-  bool valid_jobReference =
-    // Async operations are waitable.  Loosely speaking, an async operation
-    // is split into independent send and ack events, and an sync operation
-    // is not.
-    IMPLIES(  waitable, pOp->u.operation.jobReference == 2 ) &&
-    IMPLIES( !waitable, pOp->u.operation.jobReference == 1 );
+  // Async operations are waitable.  Loosely speaking, an async operation
```
Comment (markrtuttle) [inline cbmc/proofs/mqtt_state.c:113]: This refactoring seems unnecessary.  Is it semantically equivalent?  Why is this necessary?

File: cbmc/proofs/IotMqtt_Notify/Makefile (hunk 1 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Notify/Makefile:26]: It's not used, I'll remove it.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 2 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:45]: Yes, they are necessary here in their current form.

File: cbmc/proofs/IotMqtt_Notify/Makefile (hunk 1 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/Makefile:26]: Thanks.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 2 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:45]: Thanks

File: cbmc/proofs/mqtt_state.c (hunk 5 above)
Comment (markrtuttle) [reply cbmc/proofs/mqtt_state.c:113]: We discussed this.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 3 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:106]: Yes, incoming publish messages are processed by ProcessIncomingPublish (I added a comment).

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 4 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:101]: Yes, I also added a comment.

File: cbmc/proofs/mqtt_state.c (hunk 5 above)
Comment (feliperodri) [reply cbmc/proofs/mqtt_state.c:113]: I've also added a comment to clarify. Implication would be more natural here, but the use of it with CBMC assumptions leads to inconsistent values in `jobReferences`.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 4 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:103]: Sorry, but what would be wrong with a connection being destroyed?

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 3 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:109]: I think a better comment would be that "Inbound packets are either an inbound publish or an inbound response (a ping response or an acknowledgement).  The purpose of Notify is to alert any task waiting on an inbound response. Notify is never invoked on an inbound publish."

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 4 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:103]: Comment proposal: If there are no operations waiting on this connection, then there is nothing to notify, so assume references is positive.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 3 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:112]: Implemented as suggested.

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 4 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:103]: Good

File: cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c (hunk 3 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c:112]: Good.
